.buildkite/scripts/imageprep.sh

```diff
@@ -2,23 +2,47 @@
 
 set -euo pipefail
 
-source .buildkite/scripts/setenv.sh
-
-DOCKER_REGISTRY = 'docker.elastic.co'
-DOCKER_REGISTRY_SECRET_PATH = "kv/ci-shared/platform-ingest/docker_registry_prod"
+DOCKER_REGISTRY_SECRET_PATH="kv/ci-shared/platform-ingest/docker_registry_prod"
 
 publish_docker_image() {
-    echo "Pushing the docker image "$DOCKER_IMAGE:$DOCKER_IMAGE_TAG" to the ${DOCKER_REGISTRY} registry..."
-    dockerLogin
-    docker push "$DOCKER_IMAGE:$DOCKER_IMAGE_TAG"
+    echo "Pushing the docker image "$DOCKER_IMAGE":"$DOCKER_IMAGE_TAG" to the "${DOCKER_REGISTRY}" registry..."
+    docker_login
+    docker push "${DOCKER_IMAGE}":"${DOCKER_IMAGE_TAG}"
     docker logout "$DOCKER_REGISTRY"
 }
 
-
-dockerLogin() {
-    DOCKER_USER=$(retry 5 vault kv get -field user ${DOCKER_REGISTRY_SECRET_PATH})
-    DOCKER_PASSWORD=$(retry 5 vault kv get -field password ${DOCKER_REGISTRY_SECRET_PATH})
+docker_login() {
+    DOCKER_USER=$(retry 5 vault kv get -field user "${DOCKER_REGISTRY_SECRET_PATH}")
+    DOCKER_PASSWORD=$(retry 5 vault kv get -field password "${DOCKER_REGISTRY_SECRET_PATH}")
     docker login -u "${DOCKER_USER}" -p "${DOCKER_PASSWORD}" "${DOCKER_REGISTRY}" 2>/dev/null
+    trap cleanup EXIT
+}
+
+cleanup() {
+    echo "Deleting credentials..."
+    DOCKER_USER=""
+    DOCKER_PASSWORD=""
+    echo "Done."
```

Comment:
I guess environment variables could be kept. Their values are going to be lost when the script finishes. But, it could be kept :+1: 

I mean specially to run in this cleanup function the `docker logout` command. Thereby, the file generated by `docker` at `${HOME}/.docker/config.json` with the credentials/auth is removed or at least the authentication for that docker registry entry from the host/agent.

```suggestion
    echo "Deleting credentials..."
    DOCKER_USER=""
    DOCKER_PASSWORD=""
    docker logout "${DOCKER_REGISTRY}"
    echo "Done."
```

Reply:
But the `docker logout` command is already in the dpublish_docker_image func and in the build-image step as well. Could you please check it and tell me if I need to do anything else or just delete the cleanup func?

Reply:
In any of them, if any previous command fails, those `docker logout` commans won't be executed.

For instance, in `publish_docker_image` function , if docker push command fails, docker logout won't be executed since the falgs `-euo pipefail` have been set at the top of the script.

So, I would add that docker logout command there in the cleanup function to ensure it is executed always.

Maybe those others docker logout can be removed in favor of the cleanup function one ?

Reply:
Ok, I think I got your point. As it is now, `docker_login` just is executed when `build-image` is run. So the other steps could fail.

Looking at the definition of the step:
```yaml
      - label: "Build docker image"
        commands:
          - ".buildkite/scripts/imageprep.sh build-image"
          - ".buildkite/scripts/imageprep.sh push-image"
          - ".buildkite/scripts/imageprep.sh retag-and-push-image"
        agents:
          provider: "gcp"
```

I see some options here:
- Every execution of the script runs `docker_login`, and every execution of that scripts executes a cleanup trap to do a `docker logout`:
    - Every execution of imageprep.sh would be in charge of login and logout.
- The script `.buildkite/scripts/imageprep.sh` is used a helper script:
    - `imageprep.sh`  would assume that credentials are in-place.
    - You can create another script that manages the credentials required:
      ```bash
      set -euo pipefail

      source ./buildkite/scripts/imageprep.sh

      cleanup() {
        echo "Deleting credentials..."
        docker logout ${DOCKER_REGISTRY}
        echo "Done."
      }
      trap cleanup EXIT

      docker_login

      .buildkite/scripts/imageprep.sh build-image
      .buildkite/scripts/imageprep.sh push-image
      .buildkite/scripts/imageprep.sh retag-and-push-image

      docker logout ${DOCKER_REGISTRY}  # maybe this is not needed, since it is already executed by the cleanup
      ```
    - Step in the pipeline should be updated too:
      ```yaml
            - label: "Build docker image"
              commands:
                - ".buildkite/scripts/push_image.sh"
              agents:
                provider: "gcp"
      ```

Would any of those options make sense ? @sharbuz 
As a note, `build-image` probably does not require to run `docker login`. The other two (`push-image` and `retag-and-push-image`) should require to be logged in.

Reply:
I like the idea of the wrapper script.